src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Hover/RazorHoverEndpoint.cs

```diff
@@ -0,0 +1,110 @@
+        private readonly DocumentResolver _documentResolver;
+        private readonly RazorHoverInfoService _hoverInfoService;
+
+        public RazorHoverEndpoint(
+            ForegroundDispatcher foregroundDispatcher,
+            DocumentResolver documentResolver,
+            RazorHoverInfoService hoverInfoService,
+            ILoggerFactory loggerFactory)
+        {
+            if (foregroundDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (hoverInfoService is null)
+            {
+                throw new ArgumentNullException(nameof(hoverInfoService));
+            }
+
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _documentResolver = documentResolver;
+            _hoverInfoService = hoverInfoService;
+            _logger = loggerFactory.CreateLogger<RazorHoverEndpoint>();
+        }
+
+        public TextDocumentRegistrationOptions GetRegistrationOptions()
+        {
+            return new TextDocumentRegistrationOptions()
+            {
+                DocumentSelector = RazorDefaults.Selector
+            };
+        }
+
+        public async Task<HoverModel> Handle(HoverParams request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var document = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.AbsolutePath, out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler);
+
+            if(document is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var tagHelperDocumentContext = codeDocument.GetTagHelperContext();
+
+            var sourceText = await document.GetTextAsync();
+            var linePosition = new LinePosition((int)request.Position.Line, (int)request.Position.Character);
+            var hostDocumentIndex = sourceText.Lines.GetPosition(linePosition);
+            var location = new SourceSpan(hostDocumentIndex, (int)request.Position.Line, (int)request.Position.Character, 0);
+
+            var hoverItem = _hoverInfoService.GetHoverInfo(codeDocument, location);
+
+            return hoverItem;
```

Comment:
Might want add some trace logging here so we know this ran and found something. I think we do it for completion.